Automatic doc generation for major binaries

eparis opened the pull request:
Fixes #8493.

Converts apiserver, controller-manager, scheduler, proxy, kubelet to spf13/cobra. It then uses cobra to generate md, man, and bash completions. We currently do this for kubectl, this does it for more.

The one thing to note is that cobra `Run()` functions do not return an error. Thus I started using `glog.Fatal()`. This can be undone if need be...

eparis committed:
Convert kube-controller-manager to use cobra

eparis committed:
Convert kube-proxy to cobra

eparis committed:
Convert kube-scheduler to cobra

eparis committed:
Code to autogenerate markdown, man pages, and bash completions

eparis committed:
Convert kube-apiserver to cobra

eparis committed:
Convert kubelet to cobra

bgrant0607 commented:
cc @mikedanese

mikedanese commented:
Thanks @eparis! Feel free to assign to me when this is ready.

eparis commented:
It is ready, but I'd like someone like @smarterclayton or @pweil- to tell me if removing the error return to `Run()` is going to be a big pain for them?

eparis commented:
https://github.com/kubernetes/kubernetes/pull/13018/files#diff-36e1d7088209b1ab6ff878b18bb35210R298

smarterclayton commented:
Run is a big problem for us.

smarterclayton commented:
We need all the errors for most of the startup stuff - glog.Fatalf really needs to only be in main packages and the cobra stub

eparis commented:
Ok, I made no changes to `Run()` or any returns. I should point out, every single one uses `glog.Fatalf()` in some places. Someone likely will want to look into those if OpenShift doesn't like them...

smarterclayton commented:
It's not really about Openshift and code reuse as much as it is that using glog.Fatalf means that code can't be tested in a unit test.  We removed the vast majority of them from the cli, but did not clean this path.  I'll open an issue to track the rest and make reviewers aware.

bgrant0607 commented:
Is there a separate issue for returning errors from Run?

Is this still worth doing given that we're trying to eliminate all these flags and switch to configuration files?

eparis commented:
@bgrant0607 are you asking if this is worthwhile at all?  It's very little code change for consistent docs. Even if we don't want to autogenerate for kube-*, i'll likely still want to push about 80% of the code changes for cleaner generation for kubectl, which isn't going to go away...

bgrant0607 commented:
Ok

mikedanese commented:
Thanks, I'll review tomorrow or this weekend.

eparis commented:
ok to test

eparis commented:
ok to test

mikedanese commented on cmd/kube-proxy/app/server.go:
```diff
@@ -68,7 +72,7 @@ type ProxyServer struct {
 // NewProxyServer creates a new ProxyServer object with default parameters
 func NewProxyServer() *ProxyServer {
-	return &ProxyServer{
+	s := ProxyServer{
```
was this change on purpose?

eparis replied:
not really on purpose. I originally killed this func, clayton complained/said they needed it, so I restored it. But I restored it by hand, following the same convention as the other New*Server() functions. So it isn't needed, but does make the code look slightly more like the other similar code...

mikedanese commented on cmd/kube-proxy/app/server.go:
```diff
@@ -76,6 +80,38 @@ func NewProxyServer() *ProxyServer {
+
+	s.AddFlags(cmd.Flags())
+	pflag.CommandLine.AddGoFlagSet(goflag.CommandLine)
+	cmd.SetGlobalNormalizationFunc(util.WordSepNormalizeFunc)
```
Is InitFlags unused now? Can we refactor it to do what's done here and repeated in other flags? If not, can we remove it?

eparis replied:
Still used by contrib/mesos and integration. maybe I can keep working on those. Mesos really needs to make the switch to get sanity on their flags...  But separate PR

mikedanese commented on plugin/cmd/kube-scheduler/scheduler.go:
```diff
@@ -33,14 +30,11 @@ func init() {
-	verflag.PrintAndExitIfRequested()
-
-	s.Run(pflag.CommandLine.Args())
+	cmd.Execute()
```
I'm not sure what would cause a cobra command to return an error without calling s.Run but glog.Fatal(cmd.Execute()) to be safe?

mikedanese commented on hack/verify-generated-docs.sh:
```diff
@@ -22,8 +22,17 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
 source "${KUBE_ROOT}/hack/lib/init.sh"
 
 kube::golang::setup_env
+binaries=(
```
extreme nit: add a space before this line to make the diff match the diff of hack/update-generated-docs.sh

mikedanese commented:
Small nits but this looks. (this is PR 90% autogenerated). Sorry for slow review. Thanks!

eparis committed:
Do kubectl generation in kubectl instead of in 3 seperate binaries

eparis commented:
travis is 'realish'   docs are not deterministic. crap.

mikedanese commented:
LGTM once this is consistently passing tests

eparis commented:
yeah, not sure what to do. the value of GOMAXPROCS during `go build` leaks into the default value of test.parallel because of https://github.com/kubernetes/kubernetes/issues/13565  Not totally sure how to handle that...

eparis commented:
I might break this into multiple PRs and get as much in as I can.

eparis committed:
Update github.com/spf13/cobra and spf13/pflag

bump(github.com/spf13/pflag) 53d96508f5c9311f5dade714e1de7bd6a9d2af4a
bump(github.com/spf13/cobra) 5a13a75b13d132dc7d07dd19fd260257cd9c0290

This adds support for both man page generation and for hidden commands

eparis committed:
Update docs after changes

mikedanese commented:
@eparis can this be closed as it's been broken up?

mikedanese commented:
PR size appears to be nondeterministic. @lavalamp

eparis commented:
@lavalamp I'll take a look, not sure how this can happen, unless you have stuff in the logs about `.generated_docs`....

lavalamp commented:
I bet https://github.com/kubernetes/contrib/pull/100 was the cause of it switching to XL a day ago-- let's see if it does it again.

mikedanese commented:
I think this can be closed.
